zerver/views/home.py

```diff
@@ -187,6 +188,10 @@ def home_real(request: HttpRequest) -> HttpResponse:
     if need_accept_tos(user_profile):
         return accounts_accept_terms(request)
 
+    if request.session.get('needs_to_change_password', False):
+        return HttpResponseRedirect(generate_password_reset_url(user_profile,
```

Comment:
I think a comment explaining this is important. This block will be very confusing for someone reading `home_realm` code without knowing the story.

Reply:
Added.

Reply:
I applied this change to clean up the interface; we should never use `.replace()` on URLs as it can lead to security bugs:
```
diff --git a/zerver/forms.py b/zerver/forms.py
index 77b8925772..df5a6764bb 100644
--- a/zerver/forms.py
+++ b/zerver/forms.py
@@ -225,10 +225,11 @@ class LoggingSetPasswordForm(SetPasswordForm):
         return self.user
 
 def generate_password_reset_url(user_profile: UserProfile,
-                                token_generator: PasswordResetTokenGenerator) -> str:
+                                token_generator: PasswordResetTokenGenerator,
+                                url_name="password_reset_confirm") -> str:
     token = token_generator.make_token(user_profile)
     uid = urlsafe_base64_encode(force_bytes(user_profile.id))
-    endpoint = reverse('password_reset_confirm',
+    endpoint = reverse(url_name,
                        kwargs=dict(uidb64=uid, token=token))
     return f"{user_profile.realm.uri}{endpoint}"
 
diff --git a/zerver/views/home.py b/zerver/views/home.py
index 916a6a3fa6..d7e77bcc21 100644
--- a/zerver/views/home.py
+++ b/zerver/views/home.py
@@ -167,8 +167,13 @@ def home_real(request: HttpRequest) -> HttpResponse:
     # with the code in `authenticate` of EmailAuthBackend. Here, we redirect them
     # to password reset form stating that the user's password is too weak.
     if request.session.get('needs_to_change_password', False):
-        return HttpResponseRedirect(generate_password_reset_url(user_profile,
-                                    default_token_generator).replace('password', 'weak_password'))
+        return HttpResponseRedirect(
+            generate_password_reset_url(
+                user_profile,
+                default_token_generator,
+                url_name="change_weak_password"
+            )
+        )
 
     narrow, narrow_stream, narrow_topic = detect_narrowed_window(request, user_profile)
 
```